smach_ros/smach_ros/simple_action_state.py

```diff
@@ -208,76 +214,108 @@ def __init__(self,
 
         # Declare some status variables
         self._activate_time = self.node.get_clock().now()
-        self._preempt_time = self.node.get_clock().now()
         self._duration = Duration(seconds=0.0)
-        self._status = SimpleActionState.WAITING_FOR_SERVER
+        self._status = ActionState.WAITING_FOR_SERVER
 
-        # Construct action client, and wait for it to come active
+        # Construct action client and goal handle
         self._action_client = rclpy.action.ActionClient(self.node, action_spec, action_name)
+        self._client_goal_handle = None
 
         self._execution_timer_thread = None
         # Condition variables for threading synchronization
         self._done_cond = threading.Condition()
 
     def _execution_timer(self):
         """Internal method for cancelling a timed out goal after a timeout."""
-        rate = self.node.create_rate(10)
-        clock = self.node.get_clock()
-        while self._status == SimpleActionState.ACTIVE and rclpy.ok():
+        self.node.get_logger().debug(
+            f"Starting execution timer for action '{self._action_name}' "
+            f"with timeout {self._exec_timeout.nanoseconds / 1e9} seconds."
+        )
+        while self._status == ActionState.ACTIVE and rclpy.ok():
             try:
-                rate.sleep()
-            except:
-                if rclpy.ok():
-                    self.node.get_logger().error("Failed to sleep while running '%s'" % self._action_name)
-            if clock.now() - self._activate_time > self._exec_timeout:
-                self.node.get_logger().warn("Action %s timed out after %d seconds." % (self._action_name, self._exec_timeout.to_sec()))
-                # Cancel the goal
-                self._action_client.cancel_goal()
-
-    ### smach State API
+                # Check for preemption while checking for timeout
+                if self.preempt_requested():
+                    self.node.get_logger().info(f"Preempting while waiting for goal (request) result '{self._action_name}'.")
+                    self._status = ActionState.CANCELED
+                    with self._done_cond:
+                        self._done_cond.notify()
+                if self.node.get_clock().now() - self._activate_time > self._exec_timeout:
+                    self.node.get_logger().warn(
+                        f"Action {self._action_name} timed out after {self._exec_timeout.nanoseconds / 1e9} seconds. "
+                        f"Trying to cancel goal with timeout {self._preempt_timeout.nanoseconds / 1e9}.")
+                    self._status = self.__cancel_goal_and_wait(
+                        timeout=self._preempt_timeout)
+                    with self._done_cond:
+                        self._done_cond.notify()
+                time.sleep(0.1)
+            except Exception as e:
```

Comment:
Can you catch a more specific exception here?

Reply:
Not really... there are quite some possible exceptions to catch here

Reply:
I agree with @Rayman here. Catch only expected exceptions. Catching all might even be undesired at some levels.
If you don't know all, start with the one you encountered and we can expand the list later.